File: src/services/gemini/pronunciation.rs
use dioxus::prelude::*;
use crate::models::pronunciation::PronunciationEvaluation;

#[server]
pub async fn evaluate_pronunciation_server(
    language: String,
    target_sentence: String,
    spoken_transcript: String,
) -> Result<PronunciationEvaluation, ServerFnError> {
    use reqwest::Client;
    use serde_json::json;

    #[cfg(not(target_arch = "wasm32"))]
    dotenvy::dotenv().ok();

    let gemini_api_key = std::env::var("GEMINI_API_KEY")
        .map_err(|_| ServerFnError::new("Kunci GEMINI_API_KEY belum dikonfigurasi di file .env!"))?;

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(20))
        .build()
        .map_err(|e| ServerFnError::new(format!("Gagal menyiapkan HTTP client: {e}")))?;

    let gemini_model = super::model_for_chat(); 
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:generateContent?key={}",
        gemini_model, gemini_api_key
    );

    let prompt = format!(
        "Anda adalah ahli evaluasi pengucapan bahasa {language}.
Kalimat target yang seharusnya diucapkan: '{target_sentence}'
Teks Speech-to-Text hasil ucapan pengguna: '{spoken_transcript}'

Evaluasi pengucapan pengguna. STT mungkin memiliki salah ejaan jika pengucapannya salah. Jika STT kosong, berarti gagal mendengarkan.
Tentukan skor 0-100 dan berikan feedback singkat dalam bahasa Indonesia.
Beri status tiap kata dari kalimat target: 'correct', 'incorrect', atau 'missing'.
Kata-kata dalam array 'word_results' HARUS SAMA PERSIS dengan kata-kata di kalimat target secara berurutan. Abaikan tanda baca dalam field 'word'.",
        language = language,
        target_sentence = target_sentence,
        spoken_transcript = spoken_transcript
    );

    let payload = json!({
        "contents": [{ "parts": [{ "text": prompt }] }],
        "generationConfig": {
            "temperature": 0.2,
            "responseMimeType": "application/json",
            "responseSchema": {
                "type": "OBJECT",
                "properties": {
                    "score": { "type": "INTEGER" },
                    "feedback": { "type": "STRING" },
                    "word_results": {
                        "type": "ARRAY",
                        "items": {
                            "type": "OBJECT",
                            "properties": {
                                "word": { "type": "STRING" },
                                "status": { "type": "STRING" }
                            },
                            "required": ["word", "status"]
                        }
                    }
                },
                "required": ["score", "feedback", "word_results"]
            }
        }
    });

    let json_response = super::gemini_post_with_retry(&client, &url, &payload, 2).await?;

    let text_content = json_response["candidates"][0]["content"]["parts"][0]["text"]
        .as_str()
        .ok_or_else(|| ServerFnError::new("Respons API Gemini tidak berisi kandidat teks yang benar."))?;

    let cleaned_text = text_content
        .trim()
        .strip_prefix("```json")
        .unwrap_or(text_content)
        .strip_prefix("```")
        .unwrap_or(text_content)
        .strip_suffix("```")
        .unwrap_or(text_content)
        .trim();

    serde_json::from_str(cleaned_text)
        .map_err(|e| ServerFnError::new(format!("Gagal parsing respons evaluasi pronunciation: {e}")))
}

#[server]
pub async fn generate_pronunciation_sentences_server(
    language: String,
    level: String,
) -> Result<Vec<String>, ServerFnError> {
    use reqwest::Client;
    use serde_json::json;

    #[cfg(not(target_arch = "wasm32"))]
    dotenvy::dotenv().ok();

    let gemini_api_key = std::env::var("GEMINI_API_KEY")
        .map_err(|_| ServerFnError::new("Kunci GEMINI_API_KEY belum dikonfigurasi di file .env!"))?;

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(20))
        .build()
        .map_err(|e| ServerFnError::new(format!("Gagal menyiapkan HTTP client: {e}")))?;

    let gemini_model = super::model_for_chat(); 
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:generateContent?key={}",
        gemini_model, gemini_api_key
    );

    let prompt = format!(
        "Buat 5 kalimat dalam bahasa {} yang sesuai untuk level CEFR {} untuk latihan pronunciation.
Kembalikan dalam bentuk JSON array string. Jangan sertakan terjemahannya, hanya kalimat bahasa target. Panjang kalimat 4 hingga 12 kata.",
        language, level
    );

    let payload = json!({
        "contents": [{ "parts": [{ "text": prompt }] }],
        "generationConfig": {
            "temperature": 0.7,
            "responseMimeType": "application/json",
            "responseSchema": {
                "type": "ARRAY",
                "items": { "type": "STRING" }
            }
        }
    });

    let json_response = super::gemini_post_with_retry(&client, &url, &payload, 2).await?;

    let text_content = json_response["candidates"][0]["content"]["parts"][0]["text"]
        .as_str()
        .ok_or_else(|| ServerFnError::new("Respons API Gemini tidak berisi kandidat teks yang benar."))?;

    let cleaned_text = text_content
        .trim()
        .strip_prefix("```json")
        .unwrap_or(text_content)
        .strip_prefix("```")
        .unwrap_or(text_content)
        .strip_suffix("```")
        .unwrap_or(text_content)
        .trim();

    serde_json::from_str(cleaned_text)
        .map_err(|e| ServerFnError::new(format!("Gagal parsing respons kalimat pronunciation: {e}")))
}

File: lib/ai-content/pronunciation.ts
import { generateText } from "ai";
import { model } from "../ai";
import { parseAiArray, parseAiJson } from "./parse";
import type { PronunciationEvaluation } from "../types";

export function buildSentencePrompt(language: string, level: string): string {
  return [
    `Buat 5 kalimat dalam bahasa ${language} yang sesuai untuk level CEFR ${level} untuk latihan pronunciation.`,
    "Kembalikan dalam bentuk JSON array string. Jangan sertakan terjemahannya, hanya kalimat bahasa target. Panjang kalimat 4 hingga 12 kata.",
  ].join("\n");
}

export function parseSentenceArray(text: string): string[] | null {
  const arr = parseAiArray<string>(text);
  if (!arr || arr.length < 1) return null;
  if (!arr.every((s) => typeof s === "string" && s.trim().length > 0)) return null;
  return arr.map((s) => s.trim());
}

export function buildEvaluationPrompt(language: string, targetSentence: string, transcript: string): string {
  return [
    `Anda adalah ahli evaluasi pengucapan bahasa ${language}.`,
    `Kalimat target yang seharusnya diucapkan: '${targetSentence}'`,
    `Teks Speech-to-Text hasil ucapan pengguna: '${transcript}'`,
    "",
    "Evaluasi pengucapan pengguna. STT mungkin memiliki salah ejaan jika pengucapannya salah. Jika STT kosong, berarti gagal mendengarkan.",
    "Tentukan skor 0-100 dan berikan feedback singkat dalam bahasa Indonesia.",
    "Beri status tiap kata dari kalimat target: 'correct', 'incorrect', atau 'missing'.",
    "Kata-kata dalam array 'word_results' HARUS SAMA PERSIS dengan kata-kata di kalimat target secara berurutan. Abaikan tanda baca dalam field 'word'.",
    "",
    'Kembalikan HANYA JSON: {"score": number 0-100, "feedback": string, "word_results": [{"word": string, "status": "correct"|"incorrect"|"missing"}]}',
  ].join("\n");
}

const VALID_STATUS = ["correct", "incorrect", "missing"];

export function parseEvaluation(text: string): PronunciationEvaluation | null {
  const data = parseAiJson<PronunciationEvaluation>(text);
  if (!data) return null;
  if (typeof data.score !== "number" || data.score < 0 || data.score > 100) return null;
  if (typeof data.feedback !== "string" || !data.feedback.trim()) return null;
  if (!Array.isArray(data.word_results)) return null;
  if (!data.word_results.every((w) => w && typeof w.word === "string" && VALID_STATUS.includes(w.status))) return null;
  return data;
}

export async function generateSentences(params: { language: string; level: string }): Promise<string[]> {
  const { language, level } = params;
  let prompt = buildSentencePrompt(language, level);
  for (let attempt = 1; attempt <= 2; attempt++) {
    const { text } = await generateText({ model, prompt, maxOutputTokens: 8192, temperature: 0.7 });
    const sentences = parseSentenceArray(text);
    if (sentences) return sentences;
    prompt += `\n\nRespons tidak valid. Kembalikan HANYA JSON array string (minimal 1 kalimat).`;
  }
  throw new Error("Gagal menghasilkan kalimat pronunciation yang valid.");
}

export async function evaluatePronunciation(params: { language: string; targetSentence: string; transcript: string }): Promise<PronunciationEvaluation> {
  const { language, targetSentence, transcript } = params;
  let prompt = buildEvaluationPrompt(language, targetSentence, transcript);
  for (let attempt = 1; attempt <= 2; attempt++) {
    const { text } = await generateText({ model, prompt, maxOutputTokens: 8192, temperature: 0.2 });
    const evaluation = parseEvaluation(text);
    if (evaluation) return evaluation;
    prompt += `\n\nRespons tidak valid. Kembalikan HANYA JSON sesuai bentuk yang diminta.`;
  }
  throw new Error("Gagal mengevaluasi pronunciation.");
}
